contrib/udfs/src/main/java/org/apache/drill/exec/udfs/UserAgentUtils.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.drill.exec.udfs;
+
+import nl.basjes.parse.useragent.UserAgentAnalyzer;
+
+public class UserAgentUtils {
+
+  private UserAgentUtils(){}
+
+  private static UserAgentAnalyzer instance = null;
+
+  public static synchronized UserAgentAnalyzer getInstance() {
```

Comment:
it's better to use here DCL to avoid constant synchronization.

@vvysotskyi @paul-rogers it is ok to call such methods from UDFs, I recall there was some discussions in Jira about scalar replacement. Does it apply here?

Reply:
What do you mean with DCL?
I happy to change the code if needed.

Reply:
Double-checked locking but even better to use holder on demand.
Maybe this article can help - https://www.baeldung.com/java-singleton-double-checked-locking

Reply:
I know Double checked locking, I was just confused about 'DCL'.

Reply:
No problem, it would be nice if you'll use here holder on demand, looks like it suits here perfectly.

Reply:
@arina-ielchiieva, in this case, it is safe to call such methods from UDFs. Problem with scalar replacement occurs when holder is passed to the method.

Reply:
@vvysotskyi thanks for confirmation.

Reply:
Note that this is called only during the `setup()` so I expect the number of calls to this to be very small and as such not a performance problem.

Reply:
Just to clarify you are not going to implement holder on demand?

Reply:
@arina-ielchiieva 2 things: 
1. If you want it using a "holder on demand" then I'll implement it. This is your project, I'm just passing by to help on this single aspect. Your call.
2. My knowledge of the inner workings of Drill is too limited to understand how you want it. I have no clue what a "holder on demand" looks like. Sorry.

Reply:
Yes, it would be really good to use holder on demand implementation because initialization-on-demand holder is always best practice for implementing singleton pattern.

Regarding how to implement it, example how to implement it was present in the link I have posted previously: `3.2. Initialization on Demand`.

To make things quicker, I have already adopted it to your use case, please replace your class `UserAgentUtils` with new class `UserAgentAnalyzerProvider` and call `UserAgentAnalyzerProvider.getInstance()` in your UDFs code.
```
public class UserAgentAnalyzerProvider {

  public static UserAgentAnalyzer getInstance() {
    return UserAgentAnalyzerHolder.INSTANCE;
  }

  private static class UserAgentAnalyzerHolder {
    private static final UserAgentAnalyzer INSTANCE = UserAgentAnalyzer.newBuilder()
      .dropTests()
      .hideMatcherLoadStats()
      .immediateInitialization()
      .build();
  }
}
```

Reply:
Thanks! I've updated the code, works on my machine and pushed it.